exercises/python/encryption-context-complete/src/document_bucket/api.py

```diff
@@ -0,0 +1,106 @@
+from typing import Dict, Set
+
+import aws_encryption_sdk  # type: ignore
+from aws_encryption_sdk import KMSMasterKeyProvider  # type: ignore
+
+from .model import (ContextItem, ContextQuery, DocumentBundle, PointerItem,
+                    PointerQuery)
+
+
+class DocumentBucketOperations:
+    def __init__(self, bucket, table, master_key_provider: KMSMasterKeyProvider):
+        self.bucket = bucket
+        self.table = table
+        self.master_key_provider: KMSMasterKeyProvider = master_key_provider
+
+    def _write_pointer(self, item: PointerItem):
+        self.table.put_item(Item=item.to_item())
+
+    def _write_object(self, data: bytes, item: PointerItem):
```

Comment:
Should we have exponential backoff on S3 bucket operations (or any network operation for that matter)?

Reply:
I've gone back and forth on this. Part of the goal of the workshop is to be as simple and approachable as possible to help people grok what can be an intimidating subject. Every extra complexity on top of that increases grok friction, especially details that are more about interacting with S3 and DynamoDB than the Encryption SDK, and this isn't a "real" system. That said, it's also important to me that I offer Best Practices Examples.

Where I've landed at the moment is to keep the base content simple and discuss the productionizing details in an "Explorations" section of the workshop instructions. Things I'd like to cover there include retry strategies and backoff, covering eventual consistency considerations and convergence between DynamoDB and S3, extending the indexing scheme, considering hot keys, etc.

What do you think, does that seem reasonable? Or is it too egregious to leave retry out?

Reply:
That sounds reasonable. How about a comment in this version that says something like "In a production system, we'd probably use exponential retry here"

Reply:
Added explicit requirement to https://github.com/aws-samples/busy-engineers-document-bucket/issues/23